source/opt/instruction.cpp

```diff
@@ -419,6 +419,48 @@ bool Instruction::IsVulkanStorageBuffer() const {
   return false;
 }
 
+bool Instruction::IsVulkanReadOnlyStorageBuffer() const {
+  if (opcode() != spv::Op::OpTypePointer) {
+    return false;
+  }
+  Instruction* base_type =
+      context()->get_def_use_mgr()->GetDef(GetSingleWordInOperand(1));
+
+  // Unpack the optional layer of arraying.
+  if (base_type->opcode() == spv::Op::OpTypeArray ||
+      base_type->opcode() == spv::Op::OpTypeRuntimeArray) {
+    base_type = context()->get_def_use_mgr()->GetDef(
+        base_type->GetSingleWordInOperand(0));
+  }
+
+  if (base_type->opcode() != spv::Op::OpTypeStruct) {
+    return false;
+  }
+
+  spv::StorageClass storage_class =
+      spv::StorageClass(GetSingleWordInOperand(kPointerTypeStorageClassIndex));
+
+  if (storage_class != spv::StorageClass::Uniform &&
+      storage_class != spv::StorageClass::StorageBuffer) {
+    return false;
+  }
+
+  analysis::DecorationManager* decoration_mgr = context()->get_decoration_mgr();
+  uint32_t base_result_id = base_type->result_id();
+
+  if (storage_class == spv::StorageClass::Uniform &&
+      !decoration_mgr->HasDecoration(base_result_id,
+                                     spv::Decoration::BufferBlock)) {
+    return false;
+  }
+  if (storage_class == spv::StorageClass::StorageBuffer &&
+      !decoration_mgr->HasDecoration(base_result_id, spv::Decoration::Block)) {
+    return false;
+  }
+  return decoration_mgr->HasDecoration(base_result_id,
+                                       spv::Decoration::NonWritable);
```

Comment:
I'll have to double check with spec experts to know if this is correct. This decoration means

> that **this module** does not write to the memory holding the variable. It does not prevent the use of initializers on a declaration.

If the value is a storage buffer, is it still possible that some else writes to it between the two stores? In order to be valid, there would have to be some type of memory barrier. However, I believe this algorithm is simple. It does not take control flow into consideration, so we don't know if there was a memory barrier between the two.

Reply:
I believe my latest change should incorporate your suggestion.
If the OpVariable is specifically a NonWritable storage buffer, make sure anything else with the set and binding also passes `IsReadOnlyPointerShaders`